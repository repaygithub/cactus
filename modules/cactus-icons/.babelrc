{
  "presets": ["@repay/babel-preset"],
  "sourceMaps": "inline",
  "plugins": [["@babel/plugin-proposal-private-property-in-object", { "loose": true }]]
}