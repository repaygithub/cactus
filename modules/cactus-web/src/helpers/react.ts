import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
