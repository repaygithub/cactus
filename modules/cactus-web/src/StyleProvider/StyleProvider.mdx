---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
