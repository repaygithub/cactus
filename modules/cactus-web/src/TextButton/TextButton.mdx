---
name: TextButton
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import TextButton from './TextButton'
import cactusTheme from '@repay/cactus-theme'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# TextButton

### Try it out

export const code = `<TextButton>Take Some Action</TextButton>`

<LiveProvider code={code} scope={{ TextButton }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Best practices

- `TextButton` components should not be used for navigation between pages without modification.
- If you want to make a link/anchor look like a button, there are a couple of solutions, depending on whether or not you are using Typescript. However, this will modify the button's styles.
  - JSX Solution:
  ```jsx
  <TextButton as={Link} to="/path/to/something">
    Go Somewhere!
  </TextButton>
  ```
  - TSX Solution:
  ```tsx
  const TextButtonLink = Button.withComponent(Link)
  <TextButtonLink to='/path/to/something'>Go Somewhere!</TextButtonLink>
  ```
- Buttons should not be placed within paragraphs of text and instead should be in a distinct section.
- Multiple buttons in a line should be separated with a | (pipe) and extra 4px spacing

<>

<TextButton>
Button One
</TextButton>
{' | '}
<TextButton>
Button Two
</TextButton>
</>

## Basic usage

```jsx
import { TextButton } from '@repay/cactus-web'
...
// Basic usage
<TextButton>
  Basic Text Button
</TextButton>

// Call to action variant
<TextButton variant="action">
  Take Some Action
</TextButton>

// Inverse
<TextButton inverse>
  Inversed Text Button
</TextButton>

// Disabled
<TextButton disabled>
  Disabled Text Button
</TextButton>
```

## Text + Icon Button

You can combine icons with the `TextButton` component to get a text button with an icon.

Example:

```jsx
import { TextButton } from '@repay/cactus-web'
import { StatusCheck } from '@repay/cactus-icons'
...
// Text + Icon Button
<TextButton>
  <StatusCheck />Check!
</TextButton>
```

## Properties

<PropsTable of={TextButton} />
