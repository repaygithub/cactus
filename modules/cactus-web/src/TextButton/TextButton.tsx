import { border, CactusTheme, radius, textStyle } from '@repay/cactus-theme'
import PropTypes from 'prop-types'
import { css, FlattenInterpolation, ThemeProps } from 'styled-components'
import { margin, MarginProps } from 'styled-system'

import { withStyles } from '../helpers/styled'

export type TextButtonVariants = 'standard' | 'action' | 'danger' | 'dark'

interface TextButtonProps extends MarginProps {
  variant?: TextButtonVariants
  /** !important */
  disabled?: boolean
  // No inverse danger variant
  inverse?: boolean
}

type VariantMap = { [K in TextButtonVariants]: FlattenInterpolation<ThemeProps<CactusTheme>> }

const variantMap: VariantMap = {
  action: css`
    color: ${(p): string => p.theme.colors.callToAction};
  `,
  standard: css`
    color: ${(p): string => p.theme.colors.base};
  `,
  danger: css`
    color: ${(p): string => p.theme.colors.error};
  `,
  dark: css`
    color: ${(p): string => p.theme.colors.darkContrast};
  `,
}

const inverseVariantMap: VariantMap = {
  action: css`
    font-weight: 600;
    color: ${(p): string => p.theme.colors.callToAction};
  `,
  standard: css`
    color: ${(p): string => p.theme.colors.white};
  `,
  danger: css``,
  dark: css`
    color: ${(p): string => p.theme.colors.darkContrast};
  `,
}

const disabled: FlattenInterpolation<ThemeProps<CactusTheme>> = css`
  color: ${(p): string => p.theme.colors.mediumGray};
  cursor: not-allowed;
`

const variantOrDisabled = (props: TextButtonProps) => {
  const { variant, inverse } = props
  if (variant === 'danger' && inverse) {
    throw new Error('A danger variant does not exist for an inverse TextButton.')
  }
  const map = inverse ? inverseVariantMap : variantMap
  if (props.disabled) {
    return disabled
  } else if (variant !== undefined) {
    return map[variant]
  }
}

export const focusStyle = css`
  ::after {
    content: '';
    display: block;
    box-sizing: border-box;
    position: absolute;
    height: 100%;
    width: 100%;
    top: 0;
    left: 0;
    border: ${border('callToAction')};
    border-radius: ${radius(20)};
  }
`

export const TextButton = withStyles('button', {
  displayName: 'TextButton',
  transitiveProps: ['variant', 'inverse'],
  styles: [margin],
})<TextButtonProps>`
  ${textStyle('body')};
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;

  &:hover {
    ${(p): string => (!p.disabled ? 'text-decoration: underline;' : '')}
  }

  &:focus {
    ${focusStyle}
  }

  &::-moz-focus-inner {
    border: 0;
  }

  svg {
    display: inline-block;
    margin-top: -4px;
    margin-right: 4px;
  }

  ${variantOrDisabled}
`

TextButton.propTypes = {
  variant: PropTypes.oneOf(['standard', 'action', 'danger', 'dark']),
  disabled: PropTypes.bool,
  inverse: PropTypes.bool,
}

TextButton.defaultProps = {
  variant: 'standard',
  type: 'button',
}

export default TextButton
