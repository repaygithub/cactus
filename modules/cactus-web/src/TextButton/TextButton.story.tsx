import React from 'react'

import * as icons from '@repay/cactus-icons'
import { actions } from '@storybook/addon-actions'
import { boolean, select, text } from '@storybook/addon-knobs/react'
import { storiesOf } from '@storybook/react'
import DarkMode from '../storySupport/DarkMode'
import TextButton, { TextButtonVariants } from './TextButton'

const textButtonVariants: TextButtonVariants[] = ['standard', 'action', 'danger']
type IconName = keyof typeof icons
const iconNames: IconName[] = Object.keys(icons) as IconName[]
const eventLoggers = actions('onClick', 'onFocus', 'onBlur')
const textButtonStories = storiesOf('TextButton', module)
const textIconButtonStories = storiesOf('Text+Icon Button', module)

textIconButtonStories.add('Basic Usage', () => {
  const iconName: IconName = select('icon', iconNames, 'ActionsAdd')
  const Icon = icons[iconName]
  return (
    <TextButton
      variant={select('variant', textButtonVariants, 'standard')}
      disabled={boolean('disabled', false)}
      inverse={boolean('inverse', false)}
      {...eventLoggers}
    >
      <Icon />
      {text('children', 'Add')}
    </TextButton>
  )
})
