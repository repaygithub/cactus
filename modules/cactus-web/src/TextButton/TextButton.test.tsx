import * as React from 'react'

import { cleanup, fireEvent, render } from 'react-testing-library'
import { StatusCheck } from '@repay/cactus-icons'
import { StyleProvider } from '../StyleProvider/StyleProvider'
import cactusTheme from '@repay/cactus-theme'
import TextButton from './TextButton'

afterEach(cleanup)

describe('component: TextButton', () => {
  test('should default to standard variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton>Click me!</TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should render standard variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton variant="standard">Click me!</TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should render call to action variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton variant="action">Click me!</TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should render danger variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton variant="danger">Click me!</TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should render disabled variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton disabled>Click me!</TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should render inverse standard variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton variant="standard" inverse>
          Click me!
        </TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should render inverse call to action variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton variant="action" inverse>
          Click me!
        </TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should render inverse disabled variant', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton variant="standard" inverse disabled>
          Click me!
        </TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should support space props', () => {
    const textButton = render(
      <StyleProvider>
        <TextButton marginRight={5}>I have margins!</TextButton>
      </StyleProvider>
    )

    expect(textButton.asFragment()).toMatchSnapshot()
  })

  test('should trigger onClick', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(
      <StyleProvider>
        <TextButton onClick={onClick} data-testid="clicked">
          Click me!
        </TextButton>
      </StyleProvider>
    )

    fireEvent.click(getByTestId('clicked'))
    expect(onClick).toHaveBeenCalled()
  })

  test('should not trigger onClick', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(
      <StyleProvider>
        <TextButton disabled onClick={onClick} data-testid="not-clicked">
          Click me!
        </TextButton>
      </StyleProvider>
    )

    fireEvent.click(getByTestId('not-clicked'))
    expect(onClick).not.toHaveBeenCalled()
  })

  test('should render a text+icon button', () => {
    const textIconButton = render(
      <StyleProvider>
        <TextButton>
          <StatusCheck />
          Check check
        </TextButton>
      </StyleProvider>
    )

    expect(textIconButton.asFragment()).toMatchSnapshot()
  })

  test('should render a disabled text+icon button', () => {
    const textIconButton = render(
      <StyleProvider>
        <TextButton>
          <StatusCheck />
          Check check
        </TextButton>
      </StyleProvider>
    )

    expect(textIconButton.asFragment()).toMatchSnapshot()
  })
})
