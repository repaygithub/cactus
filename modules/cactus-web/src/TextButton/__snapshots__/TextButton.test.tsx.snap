// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: TextButton should default to standard variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(200,96%,11%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render a disabled text+icon button 1`] = `
<DocumentFragment>
  .c1 {
  vertical-align: middle;
}

.c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(200,96%,11%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    <svg
      class="c1"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
      />
    </svg>
    Check check
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render a text+icon button 1`] = `
<DocumentFragment>
  .c1 {
  vertical-align: middle;
}

.c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(200,96%,11%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    <svg
      class="c1"
      fill="currentcolor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
      />
    </svg>
    Check check
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render call to action variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(200,96%,35%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render danger variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(353,84%,44%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render disabled variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(0,0%,70%);
  cursor: not-allowed;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    disabled=""
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render inverse call to action variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  font-weight: 700;
  color: hsl(200,96%,35%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render inverse disabled variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(0,0%,70%);
  cursor: not-allowed;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    disabled=""
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render inverse standard variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(0,0%,100%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should render standard variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(200,96%,11%);
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`component: TextButton should support space props 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 18px;
  line-height: 28px;
  position: relative;
  border: none;
  padding: 4px;
  outline: none;
  background-color: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-sizing: border-box;
  color: hsl(200,96%,11%);
  margin-right: 24px;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus::after {
  content: '';
  display: block;
  box-sizing: border-box;
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  border: 2px solid hsl(200,96%,35%);
  border-radius: 20px;
}

.c0::-moz-focus-inner {
  border: 0;
}

.c0 svg {
  display: inline-block;
  margin-top: -4px;
  margin-right: 4px;
}

<button
    class="c0"
    type="button"
  >
    I have margins!
  </button>
</DocumentFragment>
`;
